packages/flutter/lib/src/widgets/selectable_region.dart

```diff
@@ -1593,44 +1593,80 @@ class _DirectionallyExtendCaretSelectionAction<T extends DirectionalCaretMovemen
   }
 }
 
-class _SelectableRegionContainerDelegate extends MultiSelectableSelectionContainerDelegate {
-  final Set<Selectable> _hasReceivedStartEvent = <Selectable>{};
-  final Set<Selectable> _hasReceivedEndEvent = <Selectable>{};
+/// This class manages updating multiple [Selectable] children where the
+/// [Selectable]s do not change or move around frequently.
+///
+/// It keeps track of the [Selectable]s that received start or end
+/// [SelectionEvent]s to accurately synthesize [SelectionEvent]s for the
+/// children [Selectable]s when needed.
+///
+/// When a new [SelectionEdgeUpdateEvent] is dispatched to a [Selectable], this
+/// delegate checks whether the [Selectable] has already received a selection
+/// update for each edge that currently exists, and synthesizes an event for the
+/// edges that have not yet received an update. This synthesized event is dispatched
+/// before dispatching the new event.
+///
+/// For example, if we have an existing start edge for this delegate and a
+/// [Selectable] receives an end [SelectionEdgeUpdateEvent] and it hasn't yet
+/// received a start [SelectionEdgeUpdateEvent], it synthesizes a start
+/// [SelectionEdgeUpdateEvent] and dispatches it before dispatching the original
+/// end [SelectionEdgeUpdateEvent].
+///
+/// See also:
+///
+///  * [MultiSelectableSelectionContainerDelegate], for the class that provides
+///  the main implementation details of this [SelectionContainerDelegate].
+class MultiStaticSelectableSelectionContainerDelegate extends MultiSelectableSelectionContainerDelegate {
+  /// The set of [Selectable]s that have received start events.
+  @protected
```

Comment:
ah this is for subclass to implement their selection logic. It would be good if we can provide some sort of low level selection API, for example
```dart
void updateInternalStateBeforeAndAfterSelection(child, callback) {
  ensureChildUpdated(child);
  callback();
   _hasReceivedStartEvent.add(child);
   _hasReceivedEndEvent.add(child);
}
```

and then subclass can do

```dart
SelectionResult handleSelectParagraph(event) {
 for(..) {
    updateInternalStateBeforeAndAfterSelection(selectable,() {_handleSelectParagraph} )
 }
}
```

If we can't completely hide these internal state, at least make them readonly, so that only one class is updating the state.